Unset field when cancelling timer causes event processing error

```
java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:168) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.processEvent(StreamProcessorController.java:304) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.logstreams.processor.StreamProcessorController.readNextEvent(StreamProcessorController.java:287) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
Caused by: java.lang.RuntimeException: Unexpected error occurred during RocksDB transaction.
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:168) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:152) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileEqualPrefix(ZeebeTransactionDb.java:337) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileEqualPrefix(ZeebeTransactionDb.java:306) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileEqualPrefix(TransactionalColumnFamily.java:81) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.state.TimerInstanceState.forEachTimerForElementInstance(TimerInstanceState.java:106) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.deployment.TransformingDeploymentCreateProcessor.createTimerIfTimerStartEvent(TransformingDeploymentCreateProcessor.java:98) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.deployment.TransformingDeploymentCreateProcessor.processRecord(TransformingDeploymentCreateProcessor.java:71) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:56) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedStreamProcessor$DelegatingEventProcessor.processEvent(TypedStreamProcessor.java:170) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.runInNewTransaction(ZeebeTransactionDb.java:176) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:165) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	... 8 more
Caused by: io.zeebe.msgpack.MsgpackPropertyException: Property 'repetitions' is invalid: Expected a value or default value to be specified, but has nothing
	at io.zeebe.msgpack.property.BaseProperty.resolveValue(BaseProperty.java:73) ~[zb-msgpack-value-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.msgpack.property.BaseProperty.getEncodedLength(BaseProperty.java:79) ~[zb-msgpack-value-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.msgpack.value.ObjectValue.getEncodedLength(ObjectValue.java:169) ~[zb-msgpack-value-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.msgpack.value.ObjectValue.getEncodedLength(ObjectValue.java:159) ~[zb-msgpack-value-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.msgpack.UnpackedObject.getLength(UnpackedObject.java:52) ~[zb-msgpack-value-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.copyExistingEventToBuffer(LogStreamBatchWriterImpl.java:177) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.logstreams.log.LogStreamBatchWriterImpl.done(LogStreamBatchWriterImpl.java:164) ~[zb-logstreams-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:114) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedCommandWriterImpl.appendRecord(TypedCommandWriterImpl.java:81) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedCommandWriterImpl.appendFollowUpCommand(TypedCommandWriterImpl.java:125) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.CatchEventBehavior.unsubscribeFromTimerEvent(CatchEventBehavior.java:140) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.deployment.TransformingDeploymentCreateProcessor.lambda$createTimerIfTimerStartEvent$0(TransformingDeploymentCreateProcessor.java:105) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.state.TimerInstanceState.lambda$forEachTimerForElementInstance$2(TimerInstanceState.java:109) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileEqualPrefix$8(ZeebeTransactionDb.java:312) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.visit(ZeebeTransactionDb.java:382) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.lambda$whileEqualPrefix$9(ZeebeTransactionDb.java:363) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:163) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.ensureInOpenTransaction(ZeebeTransactionDb.java:152) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileEqualPrefix(ZeebeTransactionDb.java:337) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.whileEqualPrefix(ZeebeTransactionDb.java:306) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.TransactionalColumnFamily.whileEqualPrefix(TransactionalColumnFamily.java:81) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.state.TimerInstanceState.forEachTimerForElementInstance(TimerInstanceState.java:106) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.deployment.TransformingDeploymentCreateProcessor.createTimerIfTimerStartEvent(TransformingDeploymentCreateProcessor.java:98) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.workflow.processor.deployment.TransformingDeploymentCreateProcessor.processRecord(TransformingDeploymentCreateProcessor.java:71) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedRecordProcessor.processRecord(TypedRecordProcessor.java:56) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.broker.logstreams.processor.TypedStreamProcessor$DelegatingEventProcessor.processEvent(TypedStreamProcessor.java:170) ~[zeebe-broker-core-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.runInNewTransaction(ZeebeTransactionDb.java:176) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.transaction(ZeebeTransactionDb.java:165) ~[zb-db-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	... 8 more
11:18:30.647 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [gateway-zb-actors-0] ERROR io.zeebe.gateway - Error handling gRPC request
io.grpc.StatusRuntimeException: INTERNAL: Unexpected error occurred during the request processing
	at io.grpc.Status.asRuntimeException(Status.java:521) ~[grpc-core-1.18.0.jar:1.18.0]
	at io.zeebe.gateway.EndpointManager.convertThrowable(EndpointManager.java:257) ~[zeebe-gateway-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.gateway.EndpointManager.lambda$sendRequest$2(EndpointManager.java:235) ~[zeebe-gateway-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequest$3(BrokerRequestManager.java:118) ~[zeebe-gateway-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequestInternal$6(BrokerRequestManager.java:193) ~[zeebe-gateway-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:40) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:90) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:53) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:189) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:154) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:135) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:112) [zb-util-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
Caused by: io.zeebe.transport.RequestTimeoutException: Request timed out after PT15S
	at io.zeebe.transport.impl.sender.OutgoingRequest.timeout(OutgoingRequest.java:151) ~[zb-transport-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at io.zeebe.transport.impl.sender.Sender.onTimerExpiry(Sender.java:483) ~[zb-transport-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	at org.agrona.DeadlineTimerWheel.poll(DeadlineTimerWheel.java:284) ~[agrona-0.9.33.jar:0.9.33]
	at io.zeebe.transport.impl.sender.Sender.processTimeouts(Sender.java:114) ~[zb-transport-0.16.0-SNAPSHOT.jar:0.16.0-SNAPSHOT]
	... 6 more
```

This bug is caused by an unset 'repetitions' field in the timer record when writing a Timer.CANCEL. It only happens if the timer creation and the timer cancellation are separated by a broker restart so the test should reflect that.

--- a/broker-core/src/main/java/io/zeebe/broker/exporter/record/value/TimerRecordValueImpl.java
+++ b/broker-core/src/main/java/io/zeebe/broker/exporter/record/value/TimerRecordValueImpl.java
@@ -28,18 +28,30 @@ public class TimerRecordValueImpl extends RecordValueImpl implements TimerRecord
   private final long workflowInstanceKey;
   private final long dueDate;
   private final String handlerFlowNodeId;
+  private final int repetitions;
+  private final long workflowKey;
 
   public TimerRecordValueImpl(
       ExporterObjectMapper objectMapper,
       long elementInstanceKey,
       long workflowInstanceKey,
       long dueDate,
-      String handlerFlowNodeId) {
+      String handlerFlowNodeId,
+      int repetitions,
+      long workflowKey) {
     super(objectMapper);
     this.elementInstanceKey = elementInstanceKey;
     this.dueDate = dueDate;
     this.handlerFlowNodeId = handlerFlowNodeId;
     this.workflowInstanceKey = workflowInstanceKey;
+
+    this.repetitions = repetitions;
+    this.workflowKey = workflowKey;
+  }
+
+  @Override
+  public long getWorkflowKey() {
+    return workflowKey;
   }
 
   @Override
@@ -63,8 +75,20 @@ public class TimerRecordValueImpl extends RecordValueImpl implements TimerRecord
   }
 
   @Override
+  public int getRepetitions() {
+    return repetitions;
+  }
+
+  @Override
   public int hashCode() {
-    return Objects.hash(elementInstanceKey, workflowInstanceKey, dueDate, handlerFlowNodeId);
+
+    return Objects.hash(
+        elementInstanceKey,
+        workflowInstanceKey,
+        dueDate,
+        handlerFlowNodeId,
+        workflowKey,
+        repetitions);
   }
 
   @Override
@@ -75,11 +99,15 @@ public class TimerRecordValueImpl extends RecordValueImpl implements TimerRecord
     if (o == null || getClass() != o.getClass()) {
       return false;
     }
-    final TimerRecordValueImpl that = (TimerRecordValueImpl) o;
-    return elementInstanceKey == that.elementInstanceKey
-        && workflowInstanceKey == that.workflowInstanceKey
-        && dueDate == that.dueDate
-        && Objects.equals(handlerFlowNodeId, that.handlerFlowNodeId);
+
+    final TimerRecordValueImpl other = (TimerRecordValueImpl) o;
+
+    return elementInstanceKey == other.elementInstanceKey
+        && workflowInstanceKey == other.workflowInstanceKey
+        && dueDate == other.dueDate
+        && repetitions == other.repetitions
+        && workflowKey == other.workflowKey
+        && Objects.equals(handlerFlowNodeId, other.handlerFlowNodeId);
   }
 
   @Override
@@ -89,8 +117,12 @@ public class TimerRecordValueImpl extends RecordValueImpl implements TimerRecord
         + elementInstanceKey
         + ", workflowInstanceKey="
         + workflowInstanceKey
+        + ", workflowKey="
+        + workflowKey
         + ", dueDate="
         + dueDate
+        + ", repetitions="
+        + repetitions
         + ", handlerFlowNodeId='"
         + handlerFlowNodeId
         + '\''

--- a/broker-core/src/main/java/io/zeebe/broker/exporter/stream/ExporterRecordMapper.java
+++ b/broker-core/src/main/java/io/zeebe/broker/exporter/stream/ExporterRecordMapper.java
@@ -357,7 +357,9 @@ public class ExporterRecordMapper {
         record.getElementInstanceKey(),
         record.getWorkflowInstanceKey(),
         record.getDueDate(),
-        asString(record.getHandlerNodeId()));
+        asString(record.getHandlerNodeId()),
+        record.getRepetitions(),
+        record.getWorkflowKey());
   }
 
   private VariableRecordValue ofVariableRecord(LoggedEvent event) {

--- a/broker-core/src/main/java/io/zeebe/broker/workflow/processor/CatchEventBehavior.java
+++ b/broker-core/src/main/java/io/zeebe/broker/workflow/processor/CatchEventBehavior.java
@@ -114,7 +114,7 @@ public class CatchEventBehavior {
       DirectBuffer handlerNodeId,
       Timer timer,
       TypedStreamWriter writer) {
-
+    timerRecord.reset();
     timerRecord
         .setRepetitions(timer.getRepetitions())
         .setDueDate(timer.getDueDate(ActorClock.currentTimeMillis()))
@@ -134,10 +134,12 @@ public class CatchEventBehavior {
   }
 
   public void unsubscribeFromTimerEvent(TimerInstance timer, TypedStreamWriter writer) {
+    timerRecord.reset();
     timerRecord
         .setElementInstanceKey(timer.getElementInstanceKey())
         .setWorkflowInstanceKey(timer.getWorkflowInstanceKey())
         .setDueDate(timer.getDueDate())
+        .setRepetitions(timer.getRepetitions())
         .setHandlerNodeId(timer.getHandlerNodeId())
         .setWorkflowKey(timer.getWorkflowKey());
 

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/processor/CatchEventBehaviorTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/processor/CatchEventBehaviorTest.java
@@ -1,0 +1,122 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.workflow.processor;
+
+import static org.assertj.core.api.Java6Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.eq;
+
+import io.zeebe.broker.logstreams.processor.TypedStreamWriter;
+import io.zeebe.broker.logstreams.state.ZeebeState;
+import io.zeebe.broker.subscription.command.SubscriptionCommandSender;
+import io.zeebe.broker.workflow.state.TimerInstance;
+import io.zeebe.model.bpmn.util.time.TimeDateTimer;
+import io.zeebe.msgpack.UnpackedObject;
+import io.zeebe.protocol.impl.record.value.timer.TimerRecord;
+import io.zeebe.protocol.intent.TimerIntent;
+import java.util.Random;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+public class CatchEventBehaviorTest {
+  private static final String TIME_DATE = "1990-01-01T00:00:00Z";
+
+  @Mock private ZeebeState zeebeState;
+  @Mock private SubscriptionCommandSender commandSender;
+  @Mock private TypedStreamWriter streamWriter;
+  @Captor private ArgumentCaptor<UnpackedObject> captor;
+
+  private CatchEventBehavior catchEventBehavior;
+  private Random random = new Random();
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    catchEventBehavior = new CatchEventBehavior(zeebeState, commandSender, 1);
+  }
+
+  @Test
+  public void shouldWriteCorrectTimerRecord() {
+    // given
+    final TimeDateTimer timer = TimeDateTimer.parse(TIME_DATE);
+    final long dueDate = timer.getDueDate(0);
+    final long repetitions = timer.getRepetitions();
+    final long elementInstanceKey = random.nextLong();
+    final long workflowKey = random.nextLong();
+    final long workflowInstanceKey = random.nextLong();
+
+    final byte[] buffer = new byte[5];
+    random.nextBytes(buffer);
+    final UnsafeBuffer handlerNodeId = new UnsafeBuffer(buffer);
+
+    // when
+    catchEventBehavior.subscribeToTimerEvent(
+        elementInstanceKey,
+        workflowInstanceKey,
+        workflowKey,
+        new UnsafeBuffer(handlerNodeId),
+        timer,
+        streamWriter);
+
+    // then
+    Mockito.verify(streamWriter).appendNewCommand(eq(TimerIntent.CREATE), captor.capture());
+    final TimerRecord record = (TimerRecord) captor.getValue();
+
+    assertThat(record.getHandlerNodeId()).isEqualTo(handlerNodeId);
+    assertThat(record.getElementInstanceKey()).isEqualTo(elementInstanceKey);
+    assertThat(record.getWorkflowKey()).isEqualTo(workflowKey);
+    assertThat(record.getRepetitions()).isEqualTo(repetitions);
+    assertThat(record.getDueDate()).isEqualTo(dueDate);
+  }
+
+  @Test
+  public void shouldWriteCorrectCancelTimerRecord() {
+    // given
+    final TimerInstance timer = new TimerInstance();
+    timer.setKey(random.nextLong());
+    timer.setDueDate(random.nextLong());
+    timer.setRepetitions(random.nextInt());
+    timer.setWorkflowKey(random.nextLong());
+    timer.setElementInstanceKey(random.nextLong());
+    timer.setWorkflowInstanceKey(random.nextLong());
+
+    final byte[] buffer = new byte[5];
+    random.nextBytes(buffer);
+    timer.setHandlerNodeId(new UnsafeBuffer(buffer));
+
+    // when
+    catchEventBehavior.unsubscribeFromTimerEvent(timer, streamWriter);
+
+    // then
+    Mockito.verify(streamWriter)
+        .appendFollowUpCommand(eq(timer.getKey()), eq(TimerIntent.CANCEL), captor.capture());
+    final TimerRecord record = (TimerRecord) captor.getValue();
+
+    assertThat(record.getHandlerNodeId()).isEqualTo(timer.getHandlerNodeId());
+    assertThat(record.getElementInstanceKey()).isEqualTo(timer.getElementInstanceKey());
+    assertThat(record.getWorkflowKey()).isEqualTo(timer.getWorkflowKey());
+    assertThat(record.getWorkflowInstanceKey()).isEqualTo(timer.getWorkflowInstanceKey());
+    assertThat(record.getRepetitions()).isEqualTo(timer.getRepetitions());
+    assertThat(record.getDueDate()).isEqualTo(timer.getDueDate());
+  }
+}

--- a/exporter-api/src/main/java/io/zeebe/exporter/record/value/TimerRecordValue.java
+++ b/exporter-api/src/main/java/io/zeebe/exporter/record/value/TimerRecordValue.java
@@ -25,6 +25,9 @@ import io.zeebe.protocol.intent.TimerIntent;
  */
 public interface TimerRecordValue extends RecordValue {
 
+  /** @return the key of the workflow in which this timer was created */
+  long getWorkflowKey();
+
   /** @return the key of the related element instance. */
   long getElementInstanceKey();
 
@@ -43,4 +46,7 @@ public interface TimerRecordValue extends RecordValue {
    * @return the ID of the flow node which will handle the trigger element
    */
   String getHandlerFlowNodeId();
+
+  /** @return the number of times this timer should trigger */
+  int getRepetitions();
 }

--- a/test-util/src/main/java/io/zeebe/test/util/record/TimerRecordStream.java
+++ b/test-util/src/main/java/io/zeebe/test/util/record/TimerRecordStream.java
@@ -48,4 +48,12 @@ public class TimerRecordStream extends ExporterRecordStream<TimerRecordValue, Ti
   public TimerRecordStream withHandlerNodeId(final DirectBuffer handlerNodeId) {
     return withHandlerNodeId(bufferAsString(handlerNodeId));
   }
+
+  public TimerRecordStream withRepetitions(final int repetitions) {
+    return valueFilter(v -> v.getRepetitions() == repetitions);
+  }
+
+  public TimerRecordStream withWorkflowKey(final long workflowKey) {
+    return valueFilter(v -> v.getWorkflowKey() == workflowKey);
+  }
 }

